walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,90 @@
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    HashSet<String> attendees = new HashSet<String>(request.getAttendees());
+    ArrayList<TimeRange> busyTimes = new ArrayList<>();
+    ArrayList<TimeRange> freeTime = new ArrayList<>();
+    ArrayList<TimeRange> meets = new ArrayList<>();
+    long requestDuration = request.getDuration();
+    int start = TimeRange.START_OF_DAY; 
+    int end = TimeRange.END_OF_DAY;
+    TimeRange day = TimeRange.WHOLE_DAY;
+
+    // Covering base cases of a request that is longer than the day or having no events
+    if (requestDuration>=day.duration()){
+        return meets;
+    }
+    if (events.size()==0){
+        meets.add(TimeRange.WHOLE_DAY);
+        return meets;
+    }
+
+    // Adds point of start of the day to aid in coalescing of busy chunks
+    busyTimes.add(TimeRange.fromStartEnd(start, start, false));
+
+    // Determines which events have attendees of interest in them and adds that time range to list
+    for (Event event:events){
+        if (!Collections.disjoint(event.getAttendees(),attendees)){
+            busyTimes.add(event.getWhen());
+        }
+    }
+    // Adds point of end of the day to aid in coalescing busy chunks 
+    busyTimes.add(TimeRange.fromStartEnd(end+1, end+1,false));
+
+    coalesceTimes(busyTimes);
+    findFreePeriods(freeTime, busyTimes, busyTimes.size());
+    parseMeetTimes(freeTime, meets, requestDuration);
+
+    return meets;
   }
+    // Coalesces overlapping busy chunks by setting the start to be the earlier of 2 chunks 
+    // and the end to be the later of the 2 - the separate chunks are removed from the list and 
+    // the coalesced chunk is added in its place
+    public void coalesceTimes(ArrayList<TimeRange> busyTimes){
+        
+        Collections.sort(busyTimes,TimeRange.ORDER_BY_START);
+        int busyTimesLength = busyTimes.size();
+
+        for (int i = 1; i < busyTimesLength; i++){
+            TimeRange prior = busyTimes.get(i-1);
+            TimeRange current = busyTimes.get(i);
+            if (prior.overlaps(current)){
+                TimeRange combined = TimeRange.fromStartEnd(prior.start(), Math.max(prior.end(),current.end()),false);
+                busyTimes.remove(i);
+                busyTimes.remove(i-1);            
+                busyTimes.add(i-1,combined);
+                busyTimesLength--;i--;
+            }
+        }
+        return;
+    }
+
+    // Free periods found as the difference between the end of prior busy periods and the beginning of upcoming busy periods
+    // the resultant time range is added to the list freeTime
+    public void findFreePeriods(ArrayList<TimeRange> freeTime, ArrayList<TimeRange> busyTimes,int length){
```

Comment:
We should avoid having an output parameter, which is the case here with freeTime. Here is a screenshot of our Java best practices doc since you don't have access to it: https://screenshot.googleplex.com/f3OMmZijmNV